File: algorithm/134.c
#include <stdio.h>
#include <stdlib.h>

int canCompleteCircuit(int *gas, int gasSize, int *cost, int costSize)
{
    int totalGas = 0, can = -1, j;
    for (int i = 0; i < gasSize; i++)
    {
        if ((gas[i] - cost[i]) < 0)
        {
            continue;
        }
        j = i;
        totalGas = gas[j] - cost[j];
        j++;
        j = j % gasSize;
        while (j != i && totalGas >= 0)
        {
            totalGas = totalGas + gas[j] - cost[j];
            j++;
            j = j % gasSize;
        }
        if ((totalGas >= 0) && (j == i))
        {
            can = i;
            break;
        }
    }
    return can;
}

int main(int argc, char *argv[])
{
    int gas[3] = {3, 3, 4};
    int cost[3] = {3, 4, 4};
    printf("%d\n", canCompleteCircuit(gas, 3, cost, 3));
    return 0;
}

File: go/goalgor/algor/134.go
func canCompleteCircuit(gas []int, cost []int) int {
	var j int
	n := len(gas)
	can := -1
	for i := 0; i < n; i++ {
		sum := gas[i]
		// This is the start
		if i == n-1 {
			j = 0
		} else {
			j = i + 1
		}
		for j != i && j < n {
			if j != 0 {
				sum = sum - cost[j-1]
			} else {
				sum = sum - cost[n-1]
			}
			if sum < 0 {
				break
			}
			sum = sum + gas[j]
			j++
			j = j % n
		}
		if sum >= 0 && j == i {
			if j != 0 && sum-cost[j-1] >= 0 {
				can = i
			} else if j == 0 && sum-cost[n-1] >= 0 {
				can = i
			}
		}
	}
	return can
}